libwild/src/version_script.rs

```diff
@@ -86,19 +91,44 @@ impl<'data> BasicMatchRules<'data> {
         mangled: bool,
     ) -> bool {
         // Early exit before we actually demangle the name.
-        if self.exact.is_empty() {
+        if self.exact.is_empty() && self.escaped_exact.is_empty() {
             return false;
         }
 
-        if mangled {
-            let demangled_name = lookup.get_demangled_name();
-            // The creation of UnversionedSymbolName should be relatively cheap as we construct
-            // it at most twice.
-            self.exact
-                .contains(&UnversionedSymbolName::prehashed(demangled_name.as_bytes()))
-        } else {
-            self.exact.contains(lookup.name)
+        // Check normal exact matches first
+        if !self.exact.is_empty() {
+            if mangled {
+                let demangled_name = lookup.get_demangled_name();
+                // The creation of UnversionedSymbolName should be relatively cheap as we construct
+                // it at most twice.
+                if self
+                    .exact
+                    .contains(&UnversionedSymbolName::prehashed(demangled_name.as_bytes()))
+                {
+                    return true;
+                }
+            } else if self.exact.contains(lookup.name) {
+                return true;
+            }
+        }
+
+        // Check escaped exact matches
+        if !self.escaped_exact.is_empty() {
+            let symbol_bytes = if mangled {
+                let demangled_name = lookup.get_demangled_name();
+                demangled_name.as_bytes()
+            } else {
+                lookup.name.bytes()
+            };
+
+            for &escaped_pattern in &self.escaped_exact {
+                if matches_escaped_pattern(symbol_bytes, escaped_pattern) {
```

Comment:
It looks as though every time we check if a symbol matches one of the `escaped_exact` patterns, we de-escape all the patterns. It'd likely be more efficient to just de-escape them once and store those de-escaped patterns in a `HashSet`. In fact, if you wanted, you could store the de-escaped pattern in the main `HashSet` - `exact` - although you'd need to copy the de-escaped bytes into an arena in order for the lifetimes to work out.